### PR TITLE
Add flow

### PR DESCRIPTION
**Pull request by trevorgerhardt**
Also, quite a bit of refactoring around the insanity of `get-transitive-data` 😅

### CODE REVIEW

File: lib/get-transitive-data.js [hunk 1]
```diff
@@ -1,35 +1,53 @@
...
-    journeys: []
-  }
+}: {
+  from: Point,
+  log: Function,
+  origin: Origin,
+  paths: Array<PathDescriptor>,
+  query: Query,
+  to: Point,
+  transitiveNetwork: any
```
Comment (mattwigway) [inline lib/get-transitive-data.js]: Should we at least define a TransitiveNetwork type that has all the top level members (journeys, etc) as type any so it at least prevents you from accidentally passing in totally the wrong type of object?

File: lib/get-transitive-data.js [hunk 2]
```diff
@@ -61,62 +79,38 @@ export default function getTransitiveData ({
...
       journey_name: pidx,
       segments: []
     }
 
-    const path = paths[pidx]
+    const path = commonPaths[pidx]
 
-    // bail if the path-finding algorithm gave up, this is a bug but we'd rather log a javascript error than crash the tab
+    // bail if the path-finding algorithm gave up, this is a bug but we'd rather
+    // log a javascript error than crash the tab (??)
```
Comment (mattwigway) [inline lib/get-transitive-data.js:111]: i.e. we don't want to have this throw a 'cannot read property whatever of undefined' when it tries to read a property of the (null) path.

File: lib/get-transitive-data.js [hunk 3]
```diff
@@ -209,80 +203,98 @@ export default function getTransitiveData ({
...
+  pathCounts.sort(byCount)
+
+  log(`${pathCounts.length} unique paths`)
+  if (pathCounts.length > MAX_TRANSITIVE_PATHS) {
+    log(`eliminated ${pathCounts.length - MAX_TRANSITIVE_PATHS} paths with frequency <= ${pathCounts[MAX_TRANSITIVE_PATHS].count} / ${paths.length} `)
   }
 
-  // for now this is pretty simple, sort paths by the sequence of routes they use, and choose an example, eliminating the multiple-board/transfer-stops problem
-  paths.sort((p1, p2) => {
+  let allStopPatternStopSets = pathCounts
```
Comment (mattwigway) [inline lib/get-transitive-data.js:279]: I'm not clear what a StopPatternStopSet is.

File: lib/get-transitive-data.js [hunk 4]
```diff
@@ -368,66 +395,52 @@ export function bundle ({data, log}) {
...
+function getClusterDissimilarity (c1: Array<Journey>, c2: Array<Journey>, {
+  patterns,
+  stops
+}: {
+  patterns: Array<Pattern>,
+  stops: Array<Stop>
+}): number {
+  let dissimilarity = 0
 
   for (const j1 of c1) {
```
Comment (mattwigway) [inline lib/get-transitive-data.js:443]: Ugh. Who wrote code with variables like this 🙈  (it was me). I'll refactor next week.

File: lib/get-transitive-data.js [hunk 5]
```diff
@@ -439,42 +452,58 @@ function clusterDissimilarity (c1, c2, data) {
...
+  patterns,
+  stops
+}: {
+  patterns: Array<Pattern>,
+  stops: Array<Stop>
+}): number {
+  const pat1 = patterns.find((p) => p.pattern_id === s1.pattern_id)
+  const pat2 = patterns.find((p) => p.pattern_id === s2.pattern_id)
+
+  if (!pat1 || !pat2) return Infinity
```
Comment (mattwigway) [inline lib/get-transitive-data.js:485]: What's the impetus for adding this?

File: lib/types.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  segments: Array<any>
+}
+
+export type Origin = {
+  data: Int32Array,
+  index: Int32Array,
+  nMinutes: number,
+  nStops: number,
+  radius: number,
+  x: number,
```
Comment (mattwigway) [inline lib/types.js:46]: Is there a way to restrict this to integers?

File: lib/types.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+export type Place = {
+  place_id: string,
+  place_name: ?string,
+  place_lat: number,
+  place_lon: number
+}
+
+export type Point = {
+  name?: string,
+  x: number,
```
Comment (mattwigway) [inline lib/types.js:61]: Same comment about restricting to integers.

File: lib/types.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  y: number
+}
+
+export type Query = {
+  height: number,
```
Comment (mattwigway) [inline lib/types.js:66]: Same comment about integers for all of these.

**Review by mattwigway — COMMENTED** [review]
Changes mostly look good, a few questions, and I will need to review the refactor of the pathfinding code when I have a bit more time at my disposal.

File: lib/get-transitive-data.js (hunk 1 above)
Comment (trevorgerhardt) [reply lib/get-transitive-data.js]: 👌

File: lib/get-transitive-data.js (hunk 5 above)
Comment (trevorgerhardt) [reply lib/get-transitive-data.js:485]: `find` can return `undefined` and `flow` won't let you call a method on a possibly `undefined` value

File: lib/types.js (hunk 6 above)
Comment (trevorgerhardt) [reply lib/types.js:46]: I don't believe so as that would be a runtime constraint on the value contained in a JavaScript Number. Google `flowtype integers` and let me know if you see anything promising. Remember I'm only a few days ahead on `flow`y things 😉

File: lib/get-transitive-data.js (hunk 3 above)
Comment (trevorgerhardt) [reply lib/get-transitive-data.js:279]: So we go from `PathDescriptor`s which are `[stopId, pathId]` to `StopPatternStop`s which are `[stop, pattern, stop]` and when we accumulate them by segment I call that a set 😉 

...naming things, open to suggestions! 😬

Comment (trevorgerhardt) [reply lib/get-transitive-data.js:279]: Also, these `types` will hopefully turn into things we can reuse across projects so it's a good time to debate how we should represent this kind of data!

File: lib/generate-destination-data.js [hunk 9]
```diff
@@ -14,6 +16,18 @@ export default function generateDestinationData ({
   transitiveNetwork,
   waitTimes,
   walkTimes
+}: {
+  from: Point,
+  inVehicleTravelTimes: Uint8Array,
+  log: Function,
```
Comment (evansiroky) [inline lib/generate-destination-data.js:22]: [The Function type is unsafe and should be avoided](https://flow.org/en/docs/types/functions/#toc-function-type).

File: lib/index.js [hunk 10]
```diff
@@ -1,34 +1,50 @@
...
+  stopTreesLoaded: boolean = false
+  surfaceLoaded: boolean = false
 
-  constructor ({webpack = false} = {}) {
+  constructor ({webpack = false}: {webpack: boolean} = {}) {
     super(webpack ? require.resolve('./worker-handlers') : require('./worker-handlers'))
   }
 
-  log (...args) {
+  log (...args: any[]) {
```
Comment (evansiroky) [inline lib/index.js:36]: I'm cool w/ the `any` type in this case.

File: lib/index.js [hunk 11]
```diff
@@ -38,7 +54,10 @@ export default class Browsochrones extends WebWorkerPromiseInterface {
     })
   }
 
-  createTile = (coords, done) => {
+  createTile = (
+    coords: Coords,
+    done: Function
```
Comment (evansiroky) [inline lib/index.js]: [The Function type is unsafe and should be avoided](https://flow.org/en/docs/types/functions/#toc-function-type).

File: lib/isochrone-tile.js [hunk 12]
```diff
@@ -1,42 +1,30 @@
...
   height,
   scaleFactor,
   surface,
   width,
   xoffset,
   yoffset
-}) {
-  // compiler should avoid overflow checks for xp and yp because of the < 256 condition, but prevent it from checking for pixel overflow with | 0
+}: {
+  colorScheme: Function,
```
Comment (evansiroky) [inline lib/isochrone-tile.js]: [The Function type is unsafe and should be avoided](https://flow.org/en/docs/types/functions/#toc-function-type).

File: lib/origin.js [hunk 13]
```diff
@@ -1,22 +1,26 @@
...
-
 export function create ({
   data,
-  log = console.log.bind(console.log),
+  log = console.log.bind(console),
   point
-}) {
+}: {
+  data: Int32Array,
+  log: Function,
```
Comment (evansiroky) [inline lib/origin.js:21]: I'll allow this one.  ;)

File: lib/propagation.js [hunk 14]
```diff
@@ -1,25 +1,33 @@
...
-
-import {getNonTransitTime, ITERATION_WIDTH} from './origin'
-
 export default function propagate ({
   next,
   origin,
   query,
   stopTreeCache
+}: {
+  next: Function,
```
Comment (evansiroky) [inline lib/propagation.js]: [The Function type is unsafe and should be avoided](https://flow.org/en/docs/types/functions/#toc-function-type).

File: lib/types.js [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+export type Feature = {
+  type: string,
+  geometry: {
+    type: 'string',
+    coordinates: number[][]
+  }
+}
+
+export type Grid = {
+  contains: Function,
```
Comment (evansiroky) [inline lib/types.js]: [The Function type is unsafe and should be avoided](https://flow.org/en/docs/types/functions/#toc-function-type).

File: lib/types.js [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+  journey_id: string | number,
+  journey_name: string | number,
+  segments: Segment[]
+}
+
+export type TransitiveData = {
+  journeys?: Journey[],
+  patterns: Pattern[],
+  places?: Place[],
+  routes: any[],
```
Comment (evansiroky) [inline lib/types.js]: Try to avoid `any` whenever possible.

**Review by evansiroky — CHANGES_REQUESTED** [review]
Let's try to not use wildcard types like Function, Object, any or mixed.

**Comment by trevorgerhardt** [comment]
@evansiroky & @mattwigway changes are ready for another look 👌

**Comment by trevorgerhardt** [comment]
@evansiroky I left `Function` in instances where `log` is being passed around.

File: lib/types.js [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+    type: string
+  },
+  from_stop_index?: number,
+  to_stop_index?: number,
+  pattern_id?: string,
+  type: string
+}
+
+export type Stop = {
+  stop_id: number,
```
Comment (evansiroky) [inline lib/types.js]: I think this should be a string type

File: lib/types.js [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+  stop_lat: number,
+  stop_lon: number
+}
+
+export type Pattern = {
+  pattern_id: number,
+  route_id: number,
```
Comment (evansiroky) [inline lib/types.js]: I think these should also be a string type

**Review by evansiroky — APPROVED** [review]
Flow type stuff looks good to me once those `id` fields are changed to string.  Can't give a good comment on all the logic inside.

**Pull request merged**